protocol.go

```diff
@@ -51,6 +51,14 @@ func (t tcpDialer) DialSqlConnection(ctx context.Context, c *Connector, p *msdsn
 	var ips []net.IP
 	ip := net.ParseIP(p.Host)
 	if ip == nil {
+		// if the dialer has been updated, the dialer may be proxying to a different network, and so the
+		// dialer should be used to connect so the DNS is resolved within the right network
+		if c != nil && c.Dialer != nil {
```

Comment:
I don't think we're in disagreement about the end result. I'm asking if there are multiple types of custom dialers - some that proxy to other networks and some that rely on the current DNS lookup and if we'd need to distinguish between the two. EG, your dialer is clearly of the former. Anyone using the existing model, though, is getting the DNS lookup "for free" and we risk breaking them. 
Maybe a better name for `Dialer` would have been `IPDialer` since the driver has always called it with an IP address up to now. 
Your dialer could be `HostDialer` which takes a host name.
eg
```go
h, ok := c.Dialer.(HostDialer)
```

Reply:
Sounds good! I'll work on implementing the second `HostDialer` interface

Reply:
Added :)